src/Pos/Txp/Worker.hs

```diff
@@ -0,0 +1,149 @@
+            -- the first place), we ask the first node for the tx mempool.
+            liftIO (readIORef nodesRef) >>= \case
+                []          -> return ()
+                (node:rest) -> do
+                    liftIO $ writeIORef nodesRef rest
+                    txs <- getTxMempoolInvs sendActions node
+                    requestTxs sendActions node txs
+            delay $ delayInterval
+        handler (e :: SomeException) = do
+            logWarning $ "Exception arised in queryTxsWorker: " <> show e
+            delay $ delayInterval * 2
+            action `catch` handler
+    action `catch` handler
+
+-- | A specification of how 'queryTxsWorker' will do communication:
+--
+--     * It will send a MempoolMsg
+--
+--     * It will receive some InvOrData messages (containing transactions
+--       from the mempool)
+requestTxsOuts :: OutSpecs
+requestTxsOuts =
+    toOutSpecs [ convH (Proxy @(MempoolMsg TxMsgTag))
+                       (Proxy @(InvOrData TxMsgTag TxId TxMsgContents)) ]
+
+-- | Send a MempoolMsg to a node and receive incoming 'InvMsg's with
+-- transaction IDs.
+getTxMempoolInvs
+    :: WorkMode ssc m
+    => SendActions m -> DHTNode -> m [TxId]
+getTxMempoolInvs sendActions node =
+    reifyMsgLimit (Proxy @(InvOrData TxMsgTag TxId TxMsgContents)) $
+      \(_ :: Proxy s) -> converseToNode sendActions node $ \_
+        (ConversationActions{..}::(ConversationActions
+                                  (MempoolMsg TxMsgTag)
+                                  (InvOrDataLimitedLength s TxId TxMsgTag TxMsgContents)
+                                  m)
+        ) -> do
+            let txProxy = RelayProxy :: RelayProxy TxId TxMsgTag TxMsgContents
+            send $ MempoolMsg TxMsgTag
+            let getInvs = do
+                  inv' <- recv
+                  case withLimitedLength <$> inv' of
+                      Nothing -> return []
+                      Just (Right _) -> throw UnexpectedData
+                      Just (Left inv@InvMsg{..}) -> do
+                          useful <- handleInvL txProxy inv
+                          case useful of
+                              Nothing  -> getInvs
+                              Just key -> (key:) <$> getInvs
+            getInvs
+
+-- | Request several transactions.
+requestTxs
+    :: WorkMode ssc m
+    => SendActions m -> DHTNode -> [TxId] -> m ()
+requestTxs sendActions node txIds =
+    reifyMsgLimit (Proxy @(InvOrData TxMsgTag TxId TxMsgContents)) $
+      \(_ :: Proxy s) -> converseToNode sendActions node $ \_
+        (ConversationActions{..}::(ConversationActions
+                                  (ReqMsg TxId TxMsgTag)
+                                  (InvOrDataLimitedLength s TxId TxMsgTag TxMsgContents)
+                                  m)
+        ) -> do
+            let txProxy = RelayProxy :: RelayProxy TxId TxMsgTag TxMsgContents
+            let getTxs [] = return ()
+                getTxs (id:ids) = do
```

Comment:
if `getTxs` throws exception, you will drop all collected `Inv` messages, and will be lalka, right?